File: src/org/ocremix/api/entities/remix.clj
(ns org.ocremix.api.entities.remix
  (:require [org.ocremix.api.entities :as entities]
            [org.ocremix.api.persistence :as db]
            [org.ocremix.api.persistence.remixes :as db-remixes]
            [org.ocremix.api.persistence.songs :as db-songs]
            [org.ocremix.api.util.date :as date]))

(defn- date-to-year
  [m]
  (assoc m :year (when (:year m) (date/date-to-year (:year m)))))

(defn- format-game-info
  [game]
  (-> game
      (date-to-year)
      (dissoc :name_jp :publisher :system)))

(defn- format-download-info
  [remix download-urls]
  (-> remix
      (select-keys [:size :md5 :torrent])
      (assoc :links (map :url download-urls))))

(defn- format-mixpost
  [mixpost]
  (-> mixpost
      (assoc :posted (date/format-date (:posted mixpost)))
      (assoc :evaluators (db-remixes/fetch-mixpost-evaluators (:remix_id mixpost)))
      (assoc :forum_comments (:forum_link mixpost))
      (dissoc :forum_link)
      (dissoc :remix_id)))

(defn- format-remix
  [remix]
  (let [remix-id (:id remix)
        mixpost (db-remixes/fetch-mixpost remix-id)
        artists (db-remixes/fetch-remix-artists remix-id)
        album-id (:album remix)
        album (when album-id
                (db/fetch-id-name :albums album-id))
        download-urls (db-remixes/fetch-remix-downloads remix-id)
        songs (db-remixes/fetch-remix-songs remix-id)
        composers (when (seq songs)
                    (db-songs/fetch-song-composers (map :id songs)))
        game (db/fetch-entity :games (:game remix))
        publisher (db/fetch-id-name :organizations (:publisher game))
        system (db/fetch-id-name :systems (:system game))]
    (-> remix
        (date-to-year)
        (assoc :artists artists)
        (assoc :composers composers)
        (assoc :songs songs)
        (assoc :game (format-game-info game))
        (assoc :publisher publisher)
        (assoc :system system)
        (assoc :album album)
        (assoc :download (format-download-info remix download-urls))
        (dissoc :size :md5 :torrent)
        (assoc :mixpost (format-mixpost mixpost)))))

(defn get-remix
  [remix-id]
  (let [remix (entities/get-entity-info remix-id
                                        (partial db/fetch-entity :remixes)
                                        (str "Remix ID not found: " remix-id))]
    (format-remix remix)))

File: src/OCR/ApiBundle/Model/Entity/Remix.php
<?php

namespace OCR\ApiBundle\Model\Entity;

use OCR\ApiBundle\Model\Entity;
use OCR\ApiBundle\Persistence\RemixData;
use OCR\ApiBundle\Persistence\SongData;

/**
 * Model for Remix details.
 *
 * @package OCR\ApiBundle\Model\Entity
 * @author psarando
 */
class Remix extends Entity
{
    private $data;
    private $songData;

    public function __construct($db)
    {
        parent::__construct($db);
        $this->data = new RemixData($db);
        $this->songData = new SongData($db);
    }

    public function getRemix($id)
    {
        $remix = $this->data->getEntity('remixes', $id);
        if (!empty($remix)) {
            $remix = $this->formatRemix($remix);
        }

        return $remix;
    }

    private function formatRemix($remix)
    {
        $remix["year"] = $this->dateToYear($remix["year"]);

        $game = $this->data->getEntity('games', $remix["game"]);
        $remix["game"] = array_intersect_key(
            $game,
            array_flip(array(
                "id",
                "name",
                "name_short",
                "year",
            ))
        );
        $remix["game"]["year"] = $this->dateToYear($game["year"]);
        $remix["publisher"] = $this->data->getIdName('organizations', $game["publisher"]);
        $remix["system"] = $this->data->getIdName('systems', $game["system"]);

        if (!empty($remix["album"])) {
            $remix["album"] = $this->data->getIdName('albums', $remix["album"]);
        }

        $remix["artists"] = $this->data->getRemixArtists($remix["id"]);

        $songs = $this->data->getRemixSongs($remix["id"]);
        $remix["songs"] = $songs;
        $songIds = array_map(function($song) { return $song['id']; }, $songs);
        $remix["composers"] = $this->songData->getSongComposers($songIds);

        $remix["download"] = $this->formatDownloadInfo(
            $remix,
            $this->data->getDownloadUrls($remix["id"])
        );
        unset($remix["size"]);
        unset($remix["md5"]);
        unset($remix["torrent"]);

        $mixpost = $this->data->getMixPost($remix["id"]);
        $mixpost["evaluators"] = $this->data->getMixPostEvaluators($remix["id"]);
        $remix["mixpost"] = $mixpost;

        return $remix;
    }

    private function formatDownloadInfo($remix, $downloadUrls)
    {
        $download = array_intersect_key(
            $remix,
            array_flip(array(
                "size",
                "md5",
                "torrent",
            ))
        );
        $download["links"] = array_map(
            function($download) { return $download['url']; },
            $downloadUrls
        );

        return $download;
    }
}
